TestFoundation/TestFileManager.swift

```diff
@@ -1286,89 +1311,137 @@ VIDEOS=StopgapVideos
 
         // Some of these tests will throw an NSException on Darwin which would be normally be
         // modelled by a fatalError() or other hard failure, however since most of these functions
-        // are thorwable, an NSError is thrown instead which is more useful.
+        // are throwable, an NSError is thrown instead which is more useful.
         let fm = FileManager.default
 
         XCTAssertNil(fm.homeDirectory(forUser: ""))
         XCTAssertNil(NSHomeDirectoryForUser(""))
 
         XCTAssertThrowsError(try fm.contentsOfDirectory(atPath: "")) {
-            let code = CocoaError.Code(rawValue: ($0 as? NSError)!.code)
+            let code = CocoaError.Code(rawValue: ($0 as NSError).code)
             XCTAssertEqual(code, .fileReadInvalidFileName)
         }
 
         XCTAssertNil(fm.enumerator(atPath: ""))
         XCTAssertNil(fm.subpaths(atPath: ""))
         XCTAssertThrowsError(try fm.subpathsOfDirectory(atPath: "")) {
-            let code = CocoaError.Code(rawValue: ($0 as? NSError)!.code)
+            let code = CocoaError.Code(rawValue: ($0 as NSError).code)
             XCTAssertEqual(code, .fileReadInvalidFileName)
         }
 
         XCTAssertThrowsError(try fm.createDirectory(atPath: "", withIntermediateDirectories: true)) {
-            let code = CocoaError.Code(rawValue: ($0 as? NSError)!.code)
+            let code = CocoaError.Code(rawValue: ($0 as NSError).code)
+#if os(Windows)
+            XCTAssertEqual(code, .fileNoSuchFile)
+#else
             XCTAssertEqual(code, .fileReadInvalidFileName)
+#endif
         }
         XCTAssertFalse(fm.createFile(atPath: "", contents: Data()))
         XCTAssertThrowsError(try fm.removeItem(atPath: "")) {
-            let code = CocoaError.Code(rawValue: ($0 as? NSError)!.code)
+            let code = CocoaError.Code(rawValue: ($0 as NSError).code)
             XCTAssertEqual(code, .fileReadInvalidFileName)
         }
 
         XCTAssertThrowsError(try fm.copyItem(atPath: "", toPath: "/tmp/t"))  {
-            let code = CocoaError.Code(rawValue: ($0 as? NSError)!.code)
+            let code = CocoaError.Code(rawValue: ($0 as NSError).code)
+#if os(Windows)
+            XCTAssertEqual(code, .fileReadNoSuchFile)
```

Comment:
I think this is an issue in the error reporting — if a filename is invalid, the API should return `.fileReadInvalidFileName` no matter what the host OS thinks the right error is.

Reply:
We could do that, though it would mean every time we throw because for a Win32 api call failed, we'd have to check if the error is `ERROR_FILE_NOT_FOUND` then do some checks to see if it should actually be invalid path or perhaps some other more specific error. That doesn't seem cheap to me, but it would probably only be in the exceptional case.